pep-0727.rst

```diff
@@ -84,79 +84,56 @@ like to adopt it.
 Specification
 =============
 
+The main proposal is to introduce a new function, ``typing.doc()``,
+to be used when documenting Python objects.
+This function MUST only be used within :py:class:`~typing.Annotated` annotations.
+The function takes a single string argument, ``documentation``,
+and returns an instance of ``typing.DocInfo``,
+which stores the input string unchanged.
 
-``typing.doc``
---------------
+Any tool processing ``typing.DocInfo`` objects SHOULD interpret the string as
+a docstring, and therefore SHOULD normalize whitespace
+as if ``inspect.cleandoc()`` were used.
 
-The main proposal is to have a new function ``doc()`` in the ``typing`` module.
-Even though this is not strictly related to the type annotations, it's expected
-to go in ``Annotated`` type annotations, and to interact with type annotations.
+The string passed to ``typing.doc()`` SHOULD be of the form that would be a valid docstring.
+This means that `f-strings`__ and string operations SHOULD NOT be used.
+As this cannot be enforced by the Python runtime,
+tools SHOULD NOT rely on this behaviour,
+and SHOULD exit with an error if such a prohibited string is encountered.
```

Comment:
"exit with an error" is not a useful behavior for many tools implementing this; you wouldn't want your IDE to exit if it sees an invalid docstring! In general I'm not sure we need the formal SHOULD/MUST wording that was introduced in https://github.com/python/peps/pull/3316/commits/ee92cae0faf590f36e336fb1493016dc70aa7103.

Reply:
This was me adopting the stricter side of https://github.com/python/peps/pull/3310#discussion_r1310966551, but yes that's perhaps an extreme example and a warning would be better!

The MUST/SHOULD wording was an attempt to be more precise about both what is allowed and what tools should do, which I found unclear -- I don't mind if the 'standards language' is dropped, but I do think greater specificity is needed.

A

Reply:
I think it's good to note that it's just the same rules for docstrings (I was actually not fully aware of those rules), including the mention to f-strings.

Agreed, good point, that exiting with an error would only apply to some tools.

Also, I can imagine some runtime library that extracts the data from it, and it happens to be an f-string. So, the author of the code made a mistake, but the runtime library doesn't even have a way to enforce an error, as it would just get the processed string. I don't think that library would be non-conformant if it doesn't figure out a way to raise an error there. So I would say "they are allowed to error" instead of "they have to error".

Maybe it could say something like:

>  As this cannot be enforced by the Python runtime or during static analysis, tools should not rely on this behavior, they could exit with an error or show a warning in those cases.

Also, I haven't seen MUST/SHOULD wording in other PEPs, only in IETF RFCs, citing RFC 2119 and all that stuff. As that hasn't happened or hasn't been necessary here before, I don't think it would be necessary.

---

And indeed, I should clarify some things better. I'm planning on dropping the "Additional Scenarios" section, the intention was to clarify but I think it complicates things more.

I also plan on adopting more of the wordings and suggestions from Eric Traut after this.

Reply:
About 10% of PEPs use it (``Get-ChildItem -Filter "pep-????.*" | Select-String -CaseSensitive -Pattern "(MUST|SHOULD|MAY)" | Select-Object -Unique Path``) -- so I think up to you what you'd prefer.

> As this cannot be enforced by the Python runtime or during static analysis, tools should not rely on this behavior, they could exit with an error or show a warning in those cases.

This wording sounds good, though it can be enforced during static analysis (at least some forms of SA) so perhaps a tweak to that clause?

<details>
<summary>List</summary>

```
pep-0008.txt
pep-0012.rst
pep-0244.txt
pep-0290.txt
pep-0323.txt
pep-0381.txt
pep-0396.txt
pep-0426.txt
pep-0427.txt
pep-0438.txt
pep-0440.txt
pep-0444.txt
pep-0456.txt
pep-0458.txt
pep-0459.txt
pep-0463.txt
pep-0466.txt
pep-0470.txt
pep-0474.txt
pep-0476.txt
pep-0480.txt
pep-0491.txt
pep-0493.txt
pep-0496.txt
pep-0503.txt
pep-0517.txt
pep-0561.rst
pep-0566.rst
pep-0592.rst
pep-0610.rst
pep-0621.rst
pep-0624.rst
pep-0625.rst
pep-0629.rst
pep-0631.rst
pep-0633.rst
pep-0639.rst
pep-0643.rst
pep-0650.rst
pep-0658.rst
pep-0660.rst
pep-0662.rst
pep-0665.rst
pep-0671.rst
pep-0676.rst
pep-0680.rst
pep-0685.rst
pep-0691.rst
pep-0694.rst
pep-0700.rst
pep-0701.rst
pep-0704.rst
pep-0708.rst
pep-0710.rst
pep-0711.rst
pep-0714.rst
pep-0721.rst
pep-0722.rst
pep-0723.rst
pep-0725.rst
pep-3123.txt
pep-3131.txt
pep-3149.txt
```

</details>

A

Reply:
Ah! Sorry, should have searched for those words. In that case, given it's being used already, I'm fine with anything. I would normally not use it, but if you think it makes things clearer it's fine by me.

Right, back to the original (pretty much):

> As this cannot be enforced by the Python runtime, tools should not rely on this behavior, they could exit with an error or show a warning in those cases.